[BOLT] DataAggregator support for binaries with multiple text segments

paschalis-mpeis opened the pull request:
When a binary has multiple text segments, the Size is computed as the
difference of the last address of these segments from the BaseAddress.
The base addresses of all text segments must be the same.

Background:
Larger binaries get two text segments mapped when loaded in memory.
BOLT processes only the first, which is not having a correct BaseAddress,
causing a wrong computation of a BinaryMMapInfo's size.

Consequently, BOLT wrongly thinks that many of the samples fall outside
the binary and ignores them. As a result, when used in heatmaps the
output excludes all those entries and the section hotness statistics are
wrong.

This bug is present in both the AArch64 and x86 backends.

---

This patch introduces the flag 'perf-script-events' that allows passing
perf events without BOLT having to parse them using 'perf script'.
The flag is used to pass a mock perf profile that has two memory
mappings for a mock binary that has two text segments. The size of the
mapping is updated as `parseMMapEvents` now processes all text segments.

---

Example used in unit tests:
From `/proc/<BINARY PID>/maps`, we have 2 text mappings, say A and B.

```
abc0000000-abc1000000 r-xp 011c0000 103:01 1573523 BINARY
abc2000000-abca000000 r-xp 031d0000 103:01 1573523 BINARY
```

Size of text mappings:

| Mapping |  Size  |
| ------- | ------ |
| A       | ~15MB  |
| B       | ~135MB |

---

Example on a real program:
```
2f7200000-2fabca000 r--p 00000000        bolted-binary
2fabd9000-2fe47c000 r-xp 039c9000        bolted-binary <- 1st txt segment
2fe48b000-2fe61d000 r--p 0727b000        bolted-binary
2fe62c000-2fe660000 rw-p 0740c000        bolted-binary
2fe660000-2fea4c000 rw-p 00000000
2fec00000-303dad000 r-xp 07a00000        bolted-binary <- 2nd (appears only on the bolted binary)
```

paschalis-mpeis committed:
[BOLT] DataAggregator supports binaries with multiple text segments

When a binary has multiple text segments, the Size is computed as the
difference of the last address of these segments from the BaseAddress.
The base addresses of all text segments must be the same.

Background:
Larger binaries get two text segments mapped when loaded in memory.
BOLT processes only the first, which is not having a correct BaseAddress,
causing a wrong computation of a BinaryMMapInfo's size.

Consequently, BOLT wrongly thinks that many of the samples fall outside
the binary and ignores them. As a result, when used in heatmaps the
output excludes all those entries and the section hotness statistics are
wrong.

This bug is present in both the AArch64 and x86 backends.

---

This patch introduces the flag 'perf-script-events' that allows passing
perf events without BOLT having to parse them using 'perf script'.
The flag is used to pass a mock perf profile that has two memory
mappings for a mock binary that has two text segments. The size of the
mapping is updated as `parseMMapEvents` now processes all text segments.

---

Example used in unit tests:
From `/proc/<BINARY PID>/maps`, we have 2 text mappings, say A and B.

```
abc0000000-abc1000000 r-xp 011c0000 103:01 1573523 BINARY
abc2000000-abca000000 r-xp 031d0000 103:01 1573523 BINARY
```

Size of text mappings:

| Mapping |  Size  |
| ------- | ------ |
| A       | ~15MB  |
| B       | ~135MB |

---

Example on a real program:
```
2f7200000-2fabca000 r--p 00000000        bolted-binary
2fabd9000-2fe47c000 r-xp 039c9000        bolted-binary <- 1st txt segment
2fe48b000-2fe61d000 r--p 0727b000        bolted-binary
2fe62c000-2fe660000 rw-p 0740c000        bolted-binary
2fe660000-2fea4c000 rw-p 00000000
2fec00000-303dad000 r-xp 07a00000        bolted-binary <- 2nd (appears only on the bolted binary)
```

maksfb commented on bolt/lib/Profile/DataAggregator.cpp:
```diff
@@ -2009,9 +2009,6 @@ std::error_code DataAggregator::parseMMapEvents() {
       return MI.second.PID == FileMMapInfo.second.PID;
     });
```
We no longer need the code above if we are tracking all mappings.

maksfb commented on bolt/lib/Profile/DataAggregator.cpp:
```diff
@@ -2067,7 +2064,21 @@ std::error_code DataAggregator::parseMMapEvents() {
     }
 
-    BinaryMMapInfo.insert(std::make_pair(MMapInfo.PID, MMapInfo));
+    // In some larger binaries, the loaded binary gets a a second text segment
```
Essentially, instead of storing the mapping for the first segment, you suggest to store the base address and the last address calculated as `BaseAddress + Size`, right? We will need to update the description of `MMapInfo` structure accordingly.

maksfb reviewed (commented):
Hi Paschalis, thanks for the PR. I think it applies to all binaries with multiple code segments and is a welcome fix. Can you think of the way to test it? I'm a bit confused regarding the comment about the incorrect `BaseAddress`, perhaps the test case will clarify it for me.

paschalis-mpeis committed:
BOLT fails to read correctly the size of multi-segment mmaps.

paschalis-mpeis commented:
Hi @maksfb, thanks for your review.

I went with a unit test. Consider this as a starting point and feel free to suggest more changes.

---

In short:
I added a flag `perf-script-events` to bypass BOLT's 'perf script' processing,
and I use it to supply custom mmap event entries that demonstrate the problem.

I invoke `preprocessProfile` with the mock perf profile+binary and finally check the computed size. I needed to expose a const reference to `BinaryMMapInfo` for checking the size.

The unit test ignores some errors incurred by not supplying valid 'memory events' for irrelevant processing in `preprocessProfile`.

paschalis-mpeis committed:
Attempt to fix build bot failures.

- Added to CoreTests in BUILD.gn
- Hiding DataAggregator std out/err outputs

maksfb commented on bolt/lib/Profile/DataAggregator.cpp:
```diff
@@ -464,6 +470,13 @@ void DataAggregator::filterBinaryMMapInfo() {
 int DataAggregator::prepareToParse(StringRef Name, PerfProcessInfo &Process,
                                    PerfProcessErrorCallbackTy Callback) {
+  if (!opts::ReadPerfEvents.empty()) {
+    dbgs() << "PERF2BOLT: using pre-processed perf events for '" << Name
```
```suggestion
    outs() << "PERF2BOLT: using pre-processed perf events for '" << Name
```

maksfb commented on bolt/lib/Profile/DataAggregator.cpp:
```diff
@@ -2063,12 +2067,18 @@ std::error_code DataAggregator::parseMMapEvents() {
-      } else {
-        MMapInfo.BaseAddress = *BaseAddress;
       }
+      MMapInfo.BaseAddress = *BaseAddress;
```
Do you think we need a check here that `MMapInfo.BaseAddress` is calculated consistently across multiple segment mappings?

maksfb reviewed (commented):
Thanks for adding the test!

Do you think we need to check that `BaseAddress` is calculated consistently across multiple segment mappings?

Please update comments on `struct MMapInfo` to reflect the effect of multiple segments.

paschalis-mpeis committed:
Addressing reviewers

paschalis-mpeis commented:
Thanks Maks for your review.

That is a good idea. I've added an assertion to ensure BaseAddress consistency and a unit test for it.
I have also forced-push to rebase to latest main and changed the PR title.

paschalis-mpeis committed:
No need to launch perf in the background.

Causes issues on the CI too.

paschalis-mpeis commented:
Just pointing out that some tests are failing which are not related to this change or to BOLT (it's MLIR on Windows).
But if Meta has some further tests to run, it might be a good idea (I've saw this mentioned in other PRs).

paschalis-mpeis commented:
Gentle ping. I think the PR addresses all comments. Do we need any more changes?

maksfb commented on bolt/lib/Profile/DataAggregator.cpp:
```diff
@@ -2073,12 +2078,22 @@ std::error_code DataAggregator::parseMMapEvents() {
+      assert(MMapInfo.BaseAddress == BinaryMMapInfo[MMapInfo.PID].BaseAddress &&
+             "Base address on multiple segment mappings should match");
+
+    // Update section size.
```
```suggestion
    // Update mapping size.
```

maksfb commented on bolt/lib/Profile/DataAggregator.cpp:
```diff
@@ -2073,12 +2078,22 @@ std::error_code DataAggregator::parseMMapEvents() {
+             "Base address on multiple segment mappings should match");
+
+    // Update section size.
+    uint64_t EndAddress = MMapInfo.MMapAddress + MMapInfo.Size;
+    uint64_t Size = EndAddress - BinaryMMapInfo[MMapInfo.PID].BaseAddress;
```
```suggestion
    const uint64_t EndAddress = MMapInfo.MMapAddress + MMapInfo.Size;
    const uint64_t Size = EndAddress - BinaryMMapInfo[MMapInfo.PID].BaseAddress;
```

maksfb commented on bolt/lib/Profile/DataAggregator.cpp:
```diff
@@ -95,6 +95,12 @@ cl::opt<bool> ReadPreAggregated(
     "pa", cl::desc("skip perf and read data from a pre-aggregated file format"),
     cl::cat(AggregatorCategory));
 
+cl::opt<std::string>
+    ReadPerfEvents("perf-script-events",
+                   cl::desc("skip perf event collection by supplying a "
+                            "perf-script output in a textual format"),
+                   cl::cat(AggregatorCategory));
```
If we don't expect the end user to see this option, let's make it `cl::Hidden` or `cl::ReallyHidden`. Alternatively, we can modify `DataAggregator` interface for testing since you don't invoke the option via CL.

maksfb commented on bolt/unittests/Core/MemoryMaps.cpp:
```diff
@@ -0,0 +1,142 @@
+//===- bolt/unittest/Core/MemoryMaps.cpp -------------------------------===//
```
nit: adjust the line to end at 80 chars.

maksfb reviewed (approved):
Hi Paschalis, sorry for the delay. Looks good to me, but please address a few nits before committing.

paschalis-mpeis committed:
Addressing reviewers (2)

paschalis-mpeis commented:
Thanks for your review Maks. Addressed the nits and rebased to latest main to get the unrelated tests green.


I'll be merging this in a couple of working days, given no further comments from other reviewers.

zmodem commented:
We're hitting test failures after this on Linux. From a build directory:

```
$ cmake -GNinja -DCMAKE_BUILD_TYPE=Release -DLLVM_ENABLE_ASSERTIONS=OFF -DLLVM_ENABLE_PROJECTS='clang;lld;bolt' -DLLVM_ENABLE_RUNTIMES=compiler-rt -DLLVM_TARGETS_TO_BUILD='AArch64;X86' ../llvm && ninja check-bolt
[...]
******************** TEST 'BOLT-Unit :: Core/./CoreTests/26/27' FAILED ********************
Script(shard):
--
GTEST_OUTPUT=json:/work/llvm-project/build.xx/tools/bolt/unittests/Core/./CoreTests-BOLT-Unit-989610-26-27.json GTEST_SHUFFLE=0 GTEST_TOTAL_SHARDS=27 GTEST_SHARD_INDEX=26 /work/llvm-project/build.xx/tools/bolt/unittests/Core/./CoreTests
--

Script:
--
/work/llvm-project/build.xx/tools/bolt/unittests/Core/./CoreTests --gtest_filter=AArch64/MemoryMapsTester.MultipleSegmentsMismatchedBaseAddress/0
--
/work/llvm-project/bolt/unittests/Core/MemoryMaps.cpp:139: Failure
Death test: { Error Err = DA.preprocessProfile(*BC); }
    Result: failed to die.
 Error msg:
[  DEATH   ] BOLT-WARNING: build-id will not be checked because we could not read one from input binary
[  DEATH   ] Error reading BOLT data input file: line 4, column 5: expected decimal number
[  DEATH   ] Found: name
[  DEATH   ] PERF2BOLT: failed to parse samples
[  DEATH   ] Error reading BOLT data input file: line 4, column 10: expected decimal number
[  DEATH   ] Found: name
[  DEATH   ] PERF2BOLT: failed to parse memory events: Input/output error
[  DEATH   ] 


/work/llvm-project/bolt/unittests/Core/MemoryMaps.cpp:139
Death test: { Error Err = DA.preprocessProfile(*BC); }
    Result: failed to die.
 Error msg:
[  DEATH   ] BOLT-WARNING: build-id will not be checked because we could not read one from input binary
[  DEATH   ] Error reading BOLT data input file: line 4, column 5: expected decimal number
[  DEATH   ] Found: name
[  DEATH   ] PERF2BOLT: failed to parse samples
[  DEATH   ] Error reading BOLT data input file: line 4, column 10: expected decimal number
[  DEATH   ] Found: name
[  DEATH   ] PERF2BOLT: failed to parse memory events: Input/output error
[  DEATH   ] 



********************
********************
Failed Tests (2):
  BOLT-Unit :: Core/./CoreTests/AArch64/MemoryMapsTester/MultipleSegmentsMismatchedBaseAddress/0
  BOLT-Unit :: Core/./CoreTests/X86/MemoryMapsTester/MultipleSegmentsMismatchedBaseAddress/0


Testing Time: 3.13s

Total Discovered Tests: 489
  Skipped    :   7 (1.43%)
  Unsupported:  49 (10.02%)
  Passed     : 431 (88.14%)
  Failed     :   2 (0.41%)
```

Can you please take a look? I'll prepare a revert in the meantime.

paschalis-mpeis commented:
Hi Hans, thanks for reporting this.

Are these errors from an official buildbot?
I am asking as the error reported in the earlier [comment](https://github.com/llvm/llvm-project/pull/92815#issuecomment-2499024632) seems unrelated.

zmodem commented:
No, the error is from a local build. Agreed that the comment about https://lab.llvm.org/buildbot/#/builders/146/builds/1688 seems unrelated.
